### DISCUSSION TITLE
Why can't my simulation converge?

### DISCUSSION THREAD

**Discussion by abc-hy** [idaholab/moose#21826] [Q&A Modules: Phase field] [answered]
Hi everyone,

Why can't my steady state simulation converge?
Is this because of the solve type?
The following is  my input file.

```
# This simulates general particle growth kinetics
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 10002
  xmax = 6
  uniform_refine = 3
[]

[Variables]
  [./c]
  [../]
  [./w]
  [../]
[]
# aux varaibles to track the free energy change (must decrease with time)
[AuxVariables]
  [./total_F]
    order = CONSTANT
    family = MONOMIAL
  [../]
  # the chemical potential gradients
[]

[ICs]
  [./IC_c]
    x1=0
    y1=0
    x2=3
    y2=0
    inside = 0.998079634820802
    outside = -0.007983861007389 # Matrix is supersaturated with solute atoms
    variable = c
    type = BoundingBoxIC
  [../]
[]


#[BCs]
  #[./Periodic]
    #[./cx]
      #variable = c
      #auto_direction = 'x'
    #[../]
    #[./wx]
      #variable = w
      #auto_direction = 'x'
    #[../]
  #[../]

#[]



[Kernels]
  # Split form of Cahn-Hilliard equation with eta as coupled variable
  # w is the chemical potential
  #[./c_dot]
    #type = CoupledTimeDerivative
    #variable = w
    #v = c
  #[../]
  [./c_res]
    type = SplitCHParsed
    variable = c
    f_name = F
    kappa_name = kappa_c
    w = w
  [../]

  [./w_res]
    type = SplitCHWRes
    args = 'c'
    variable = w
    mob_name = M
  [../]

[]

[AuxKernels]
  [./total_F]
    type = TotalFreeEnergy
    variable = total_F
    interfacial_vars = c
    kappa_names = kappa_c
  [../]
[]




[Materials]

  [./mobility]
    type = DerivativeParsedMaterial
    f_name = M
    args = 'c'
    function = 1
    derivative_order = 1
  [../]

  [./kappa_c]
    type = GenericConstantMaterial
    prop_names  = 'kappa_c'
    prop_values = '8.0e-4'
  [../]

  [./free_energy]
    # equivalent to `MathFreeEnergy`
    type = DerivativeParsedMaterial
    f_name = F
    args = 'c'
    function = 3.222660694519139*c^2*(1.937939479530022-3.968969739765011*c^2+2*c^4)+0.1
  [../]



[]






[Postprocessors]

  [./ElementInt_c]
    type = ElementIntegralVariablePostprocessor
    variable = c
  [../]

  [./total_F]
    type = ElementIntegralVariablePostprocessor
    variable = total_F
  [../]

  [./average]
    type = ElementAverageValue
    variable = total_F
  [../]
[]
#
# [VectorPostprocessors]
#  # The numerical values of the variables/auxvariables across the centerline
#  [./line_values]
#    type =  LineValueSampler
#    start_point = '0 0 0'
#    end_point = '6 0 0'
#    variable = 'c'
#    num_points = 361
#    sort_by = x
#    #execute_on = 'FINAL'
#  [../]
# []


 ##[VectorPostprocessors]
  # The numerical values of the variables/auxvariables across the centerline
  ##[./line_values]
   ##type =  LineValueSampler
    ##start_point = '0'
    #end_point = '9'
    #variable = 'c w j_tot'
    #num_points = 199
    #sort_by =  id
    #execute_on = 'TIMESTEP_END'
  #[../]
#[]
[Preconditioning]
  [./SMP] # to produce the complete perfect Jacobian
    type = SMP
    full = true
  [../]
[]

[Executioner]

  type = Steady
  #scheme = 'BDF2'
  #petsc_options = '-snes_mf'

  #Preconditioned JFNK (default)
  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm      31                  preonly       lu           1'
  #petsc_options_iname = '-pc_type'
  #petsc_options_value = 'lu'

  l_max_its = 100 # maximum linear iterations
  l_tol = 1.0e-6 # 0.001 Linear Tolerance

  nl_max_its = 500 # maximum number of nonlinear iterations. exceed will cut dt.
  nl_rel_tol = 1.0e-9 # -8 nonlinear relative tolerance
  #num_steps = 2
  #[./TimeStepper]
    #type = IterationAdaptiveDT
    #dt = 1e-6
    #cutback_factor = 0.67
    #growth_factor = 1.5   # 1.6
    #optimal_iterations = 20
    #iteration_window = 2
  #[../]


  [./Adaptivity]
    coarsen_fraction = 0.1
    refine_fraction = 0.7
    max_h_level = 3
    interval = 1
  [../]


  # [./Adaptivity]
  #   coarsen_fraction = 0.1
  #   refine_fraction = 0.7
  #   max_h_level = 3
  #   interval = 1
  #
  #
  #
  # [../]
[]

[Outputs]
  exodus = true
  csv = true
  interval = 1

[]

```

Thank you.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Please attach the log for the run.
Also please try the solutions listed on this page https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Guillaume

**Reply by abc-hy** [reply]
what is the log for the run?

**Comment by abc-hy** [comment]
what is the log for the run?

**Reply by GiudGiud** [reply]
the console output to the screen. Please pipe it to a file then attach (upload) it here

**Reply by abc-hy** [reply]
[simulation_information.docx](https://github.com/idaholab/moose/files/9306109/simulation_information.docx)

**Reply by abc-hy** [reply]
[input_file.docx](https://github.com/idaholab/moose/files/9306116/input_file.docx)
This is the input file.

**Reply by GiudGiud** [reply]
besides the advice on that page I linked, please also look at 
- using automatic_scaling
- adding debug output to show which variable is holding up the residual

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/184144819-5337b125-31e0-43d3-be48-fc8954840c48.png)
I used Adaptivity, is this the automatic_scaling you refer to?
What is debug output?

**Reply by GiudGiud** [reply]
automatic scaling
https://mooseframework.inl.gov/newsletter/2019_07.html
debug output
https://mooseframework.inl.gov/source/actions/SetupDebugAction.html

**Reply by abc-hy** [reply]
nl_rel_tol I used is 1e(-9), is this too small or OK? Do I need to increase it?

**Reply by GiudGiud** [reply]
Depends on what you see in the log. For some problems it s too big for others it s reasonable

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/184150150-91371b03-aa82-4c87-a6f3-9b9d2415ec19.png)
Which one should I check in debug block?
![image](https://user-images.githubusercontent.com/63679626/184150247-1ab5b7c2-7c16-4c06-b795-af14b81d0b76.png)
Do I need to check material property?

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/184152675-6fa02def-79e5-469d-9cb8-a6f163c462b1.png)
Or this one?

**Reply by abc-hy** [reply]
'Debug/show_var_residual' does not hold a vec-typed value (val='true')
How to use shoe_var_residual?

**Reply by GiudGiud** [reply]
just set show_var_residual_norms to get all of them

show_var_residual is for specifying individual variables

**Reply by abc-hy** [reply]
[simulation_information2.docx](https://github.com/idaholab/moose/files/9309675/simulation_information2.docx)
This is the information of the simulation I got. What should I change?

**Reply by GiudGiud** [reply]
residual of 0 on w is suspicious. You should investigate that.

**Reply by abc-hy** [reply]
But it is in steady state.

**Reply by abc-hy** [reply]
Maybe in steady state the w will be 0?

**Reply by abc-hy** [reply]
I used the same input file, but the different initial concentration for the steady state simulation. I got the converged results, why can't I get converged results for this one?
I attached the simulation information I got by using the same code but with the different initial concentration
[simulation_information_3.docx](https://github.com/idaholab/moose/files/9311681/simulation_information_3.docx)

**Reply by abc-hy** [reply]
Do I need to change the solver option? If I change solver option, which preconditioner should I use?

**Reply by GiudGiud** [reply]
Please follow the instructions on the pages i linked earlier and report back for everything you tried. 
Examine the solution as much as you can to get an understanding of the convergence issues

**Reply by abc-hy** [reply]
I have read the troubleshooting webpage, but none of the issue listed there is similar to my case. 
I attached my simulation information as below:
[simulation Information4.docx](https://github.com/idaholab/moose/files/9332093/simulation.Information4.docx)
[input_file2.docx](https://github.com/idaholab/moose/files/9332097/input_file2.docx)
Also, i attached the changed input file, in the input file, I have already increased the number of linear iteration and number of non-linear iteration to 2000, however, as I see, in the simulation information word document, the simulation still can only run 108 non-linear iterations, what determines the number of non-linear iterations the simulation can run? Maybe there is one parameter that exceed one limit, but I don't know which parameter it is. Do you know?

**Reply by GiudGiud** [reply]
Try lu preconditioning

if that doesnt work, start from a CH input file in the repo, and modify it step by step to have the same simulation as yours. Keep track of when it stops converging

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/184520847-36506b1e-dff2-4bcc-8aad-6d1fa6e12d42.png)
Do you mean this?
There is no lu in precondition block.

What is repo?

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/184521029-798ba157-f57b-4e95-80f0-3f468d8cc9c7.png)
Do you mean this?
petsc_options_value = lu

**Reply by abc-hy** [reply]
I changed to this, but it is not right

**Reply by GiudGiud** [reply]
change it to 'lu'

repo is the short for repository. We have examples of solving the CH equations, start from there

**Reply by abc-hy** [reply]
can you provide me the link of repo?

**Reply by GiudGiud** [reply]
it's the moose repo

**Reply by abc-hy** [reply]
I changed but still does not work

**Reply by abc-hy** [reply]
Where is MOOSE repo?

**Reply by abc-hy** [reply]
https://github.com/moose/Moose
This one?
How to use it? Which one is suitable for my case?

**Reply by abc-hy** [reply]
Hi, could you run the input_file2, I sent you yesterday, on your machine? Maybe it can work on other machine? I also attached in this email again.
[input_file2.docx](https://github.com/idaholab/moose/files/9333910/input_file2.docx)

**Reply by GiudGiud** [reply]
look in the phase_field module for CH examples.
